sql/src/main/java/io/crate/expression/reference/doc/lucene/DocCollectorExpression.java

```diff
@@ -99,7 +98,34 @@ public Object value() {
             // for example:
             //      sourceExtractor might read byte as int and
             //      then eq(byte, byte) would get eq(byte, int) and fail
-            return returnType.value(sourceLookup.extractValue(columnName));
+            return returnType.value(extractValue(sourceLookup, 0, path));
+        }
+
+        @SuppressWarnings("unchecked")
+        private static Object extractValue(final Map map, int index, List<String> path) {
```

Comment:
as this is a private method can we maybe use the SourceLookup interface instead of Map? When I see a method with a Map and an index I immediately think of that index being a key (or for a stable collection an actual index)
Also, the index is in the signature is used only for the recursion but it's also confusing as to it's role in the method. Can we avoid exposing it? (especially if we drop the recursion)

Reply:
It's not really possible to change it to use the SourceLookup instead of a Map. See the second commit in which it operates on the source map directly.

I don't see an easy way to avoid the recursion. And it shouldn't go to deep as it's only 1 call per object nesting *if* that is an array. The prior implementation also used recursion.

Not sure about the arguments. I can swap the order to `map, path, index`, and maybe rename `index` to `pathStartIndex` if that helps

Reply:
You're right. Let's please just rename the index and move it after path and that'll make it a bit clearer.